File: Editorial/0103_Binary_Tree_Zigzag_Level_Order_Traversal/Go/1_BFS.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *ListNode
 *     Right *ListNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
    if root == nil {
        return nil
    }
    var res [][]int
    nodeQueue := []*TreeNode{root, nil}
    levelList := []int{}
    isOrderLeft := true
    for len(nodeQueue) > 0 {
        currNode := nodeQueue[0]
        nodeQueue = nodeQueue[1:]
        if currNode != nil {
            if isOrderLeft {
                levelList = append(levelList, currNode.Val)
            } else {
                levelList = append([]int{currNode.Val}, levelList...)
            }
            if currNode.Left != nil {
                nodeQueue = append(nodeQueue, currNode.Left)
            }
            if currNode.Right != nil {
                nodeQueue = append(nodeQueue, currNode.Right)
            }
        } else {
            res = append(res, levelList)
            levelList = []int{}
            if len(nodeQueue) > 0 {
                nodeQueue = append(nodeQueue, nil)
            }
            isOrderLeft = !isOrderLeft
        }
    }
    return res
}

File: Editorial/0101-0200/0103_Binary_Tree_Zigzag_Level_Order_Traversal/JavaScript/1_BFS.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
    if (root === null) return [];
    const results = [];
    const node_queue = [root, null];
    const level_list = [];
    let is_order_left = true;
    while (node_queue.length > 0) {
        const curr_node = node_queue.shift();
        if (curr_node !== null) {
            if (is_order_left) level_list.push(curr_node.val);
            else level_list.unshift(curr_node.val);
            if (curr_node.left !== null) node_queue.push(curr_node.left);
            if (curr_node.right !== null) node_queue.push(curr_node.right);
        } else {
            results.push([...level_list]);
            level_list.length = 0;
            if (node_queue.length > 0) node_queue.push(null);
            is_order_left = !is_order_left;
        }
    }
    return results;
};
